server/src/main/java/org/opensearch/indices/replication/checkpoint/RemoteStorePublishMergedSegmentAction.java

```diff
@@ -0,0 +1,188 @@
+                mergedSegmentCheckpoint,
+                localToRemoteStoreFilenames
+            );
+            doPublish(
+                indexShard,
+                remoteStoreMergedSegmentCheckpoint,
+                new RemoteStorePublishMergedSegmentRequest(remoteStoreMergedSegmentCheckpoint),
+                "segrep_remote_publish_merged_segment",
+                true,
+                TimeValue.timeValueMillis(timeLeftMillis)
+            );
+        } else {
+            logger.warn(
+                () -> new ParameterizedMessage(
+                    "Unable to confirm upload of merged segment {} to remote store. Timeout of {}s exceeded. Skipping pre-copy.",
+                    mergedSegmentCheckpoint,
+                    TimeValue.timeValueMillis(elapsedTimeMillis).toHumanReadableString(3)
+                )
+            );
+        }
+    }
+
+    private Map<String, String> uploadMergedSegmentsToRemoteStore(IndexShard indexShard, MergedSegmentCheckpoint checkpoint) {
+        Collection<String> segmentsToUpload = checkpoint.getMetadataMap().keySet();
+        Map<String, String> localToRemoteStoreFilenames = new ConcurrentHashMap<>();
+
+        Map<String, Long> segmentsSizeMap = checkpoint.getMetadataMap()
+            .entrySet()
+            .stream()
+            .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().length()));
+        final CountDownLatch latch = new CountDownLatch(1);
+        getRemoteStoreUploaderService(indexShard).uploadSegments(segmentsToUpload, segmentsSizeMap, new ActionListener<>() {
+            @Override
+            public void onResponse(Void unused) {
+                logger.trace(() -> new ParameterizedMessage("Successfully uploaded segments {} to remote store", segmentsToUpload));
+                latch.countDown();
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.warn(() -> new ParameterizedMessage("Failed to upload segments {} to remote store. {}", segmentsToUpload, e));
+                latch.countDown();
+            }
+        }, (x) -> new UploadListener() {
+            @Override
+            public void beforeUpload(String file) {}
+
+            @Override
+            public void onSuccess(String file) {
+                localToRemoteStoreFilenames.put(file, indexShard.getRemoteDirectory().getExistingRemoteFilename(file));
+            }
+
+            @Override
+            public void onFailure(String file) {
+                logger.warn("Unable to upload segments during merge. Continuing.");
+            }
+        }, true);
+        try {
+            long timeout = indexShard.getRecoverySettings().getMergedSegmentReplicationTimeout().seconds();
+            if (latch.await(timeout, TimeUnit.SECONDS) == false) {
+                logger.warn("Unable to confirm successful merge segment downloads by replicas. Continuing.");
```

Comment:
Lets reword the log to indicate that this is a timeout. The next log line is also identical. Also, can you add the exception in the warn logs for the 2nd exception.